Set socket options for peer connections.

batiati opened the pull request:
Socket options such as keepalive, timeout, no_delay, and send/receive buffer size were previously set only during `listen()`, leaving clients and replica-to-replica sockets with default OS parameters.

batiati committed:
Set socket options for peer connections.
Socket options such as keepalive, timeout, no_delay, and send/receive buffer size were previously set only during `listen()`, leaving clients and replica-to-replica sockets with default OS parameters.

matklad commented on src/message_bus.zig:
```diff
@@ -212,6 +212,16 @@ fn MessageBusType(comptime process_type: vsr.ProcessType) type {
+            fd: IO.socket_t,
+            address: std.net.Address,
+        } {
+            const fd = try init_tcp(io, address.any.family);
```
Given that we've extracted init_tcp, we can now pass `fd` as a parameter, and simply the signature here to return just address.

Also, `errdefer close(fd)` is needed.

matklad commented on src/message_bus.zig:
```diff
@@ -188,18 +191,15 @@ fn MessageBusType(comptime process_type: vsr.ProcessType) type {
             errdefer io.close_socket(fd);
 
-            const address_resolved = try io.listen(fd, address, .{
+            try io.socket_options(fd, .{
```
Instead of separate `io.socket_options` call, consider adding `options` parameter to `io.open_socket` so that, from the perspective of the caller, we atomically get the `fd` we want.

batiati replied:
We can't do that because some options are TCP-specific, and we use `io.open_socket` for UDP connections.
Renamed `socket_options` to `tcp_options` instead, so we can add `udp_options` when/if required.

matklad replied:
Hm, maybe we can `open_socket_tcp` / `open_socket_udp`, so that we also don't have to pass `posix.IPPROTO.TCP` which feels like it's at the wrong abstraction layer here?

matklad commented:
Excellent catch! When moving this options to message bus, I was wondering whether its enough to set them when we `.listen` or whether setting them after `.accept` is also required. I convinced myself that doing it on `.listen` is enough, but of course we have the entire other side of connection, where we certainly need also need the options!

batiati committed:
Review:
- Simplify the message bus listener API.
- Make it explicit that the options are TCP-specific, so other protocols (e.g., UDP) can implement their own settings.
- Reuse `common.setsockopt`.

batiati committed:
Review:
Split `open_socket` into `open_socket_tcp` and `open_socket_udp`.

matklad commented:
@batiati do we want to get this into Monday's release? Can do a careful review if we want, otherwise I will look with Monday eyes, don't fully trust weekend matklad!

batiati commented:
> @batiati do we want to get this into Monday's release? Can do a careful review if we want, otherwise I will look with Monday eyes, don't fully trust weekend matklad!

Sure, there is no need to rush this fix! It's raining so much in here that I'm forced to stare at the computer!😬☔️
Enjoy your weekend!

matklad commented on src/message_bus.zig:
```diff
@@ -454,8 +447,7 @@ fn MessageBusType(comptime process_type: vsr.ProcessType) type {
                 // The first replica's network address family determines the
                 // family for all other replicas:
                 const family = bus.configuration[0].any.family;
```
```suggestion
                const family = bus.configuration[replica].any.family;
```

matklad commented on src/message_bus.zig:
```diff
@@ -1,7 +1,6 @@
 const std = @import("std");
 const assert = std.debug.assert;
 const mem = std.mem;
-const posix = std.posix;
```
🎉

matklad commented on src/io/linux.zig:
```diff
@@ -1454,10 +1454,27 @@ pub const IO = struct {
 
-    /// Creates a socket that can be used for async operations with the IO instance.
-    pub fn open_socket(self: *IO, family: u32, sock_type: u32, protocol: u32) !socket_t {
+    /// Creates a UDP socket that can be used for async operations with the IO instance.
```
```suggestion
    /// Creates a TCP socket that can be used for async operations with the IO instance.
```

matklad commented on src/io/common.zig:
```diff
@@ -7,16 +7,19 @@ const assert = std.debug.assert;
+    } = null,
+    user_timeout_ms: c_int = 0,
+    nodelay: bool = false,
+};
```
FWIW, originally I didn't add defaults there by design, as per our 🐯-style:

> Explicitly pass options to library functions at the call site, instead of relying on the defaults. For example, write `@prefetch(a, .{ .cache = .data, .rw = .read, .locality = 3 });` over `@prefetch(a, .{});`. This improves readability but most of all avoids latent, potentially catastrophic bugs in case the library ever changes its defaults.

For tests, we can have `count test_options: TCPOptions = ...` to avoid verbosity.

matklad commented on src/io/common.zig:
```diff
@@ -25,72 +28,78 @@ pub fn listen(
-
-    return address_resolved;
+pub fn setsockopt(_fd: posix.socket_t, level: i32, option: u32, value: c_int) !void {
+    try posix.setsockopt(_fd, level, option, &std.mem.toBytes(value));
```
```suggestion
pub fn setsockopt(fd: posix.socket_t, level: i32, option: u32, value: c_int) !void {
    try posix.setsockopt(fd, level, option, &std.mem.toBytes(value));
```

batiati committed:
Review

matklad commented on src/io/linux.zig:
```diff
@@ -19,6 +19,7 @@ const parse_dirty_semver = stdx.parse_dirty_semver;
 const maybe = stdx.maybe;
 
 pub const IO = struct {
+    pub const TCPOptions = common.TCPOptions;
```
🤌
